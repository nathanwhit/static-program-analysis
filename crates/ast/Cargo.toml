[package]
name = "ast"
version.workspace = true
edition.workspace = true

[dependencies]
common.workspace = true
paste.workspace = true
serde = { workspace = true, optional = true }

[dev-dependencies]
ron = "0.8.0"
serde.workspace = true
pretty_assertions.workspace = true
common = { workspace = true, features = ["serde"] }

[features]
serde = ["dep:serde", "common/serde"]
