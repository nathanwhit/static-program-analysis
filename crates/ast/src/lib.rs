extern crate self as ast;

pub use print::Print;

use core::fmt;

use common::{derive_common, Foldable, SuperFoldable};
pub mod fold;
pub mod print;
pub mod visit;

pub type ExpBox = Box<Exp>;

#[derive(Copy)]
#[derive_common]
pub struct Span(pub usize, pub usize);

#[derive(Copy)]
#[derive_common]
/// Binary operator
pub enum Op {
    Add,
    Sub,
    Mul,
    Div,
    Gt,
    Eq,
    NEq,
    Lt,
}

#[derive_common]
pub struct Id(pub String);

impl fmt::Display for Id {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<String> for Id {
    fn from(s: String) -> Self {
        Id(s)
    }
}

impl From<&str> for Id {
    fn from(s: &str) -> Self {
        Id(s.to_string())
    }
}

#[derive_common]
#[derive(SuperFoldable)]
/// Expression
pub enum Exp {
    /// Integer
    Int(#[fold(identity)] i64),
    /// Identifier
    Id(Id),
    /// (Exp)
    Paren(ExpBox),
    /// Exp Op Exp
    BinOp(ExpBox, #[fold(identity)] Op, ExpBox),
    /// input
    Input,
    /// fun(Exp, Exp, ...)
    Call(ExpBox, Vec<Exp>),
    /// alloc Exp
    Alloc(ExpBox),
    /// &Id
    Ref(Id),
    /// *Exp
    Deref(ExpBox),
    /// { Id: Exp, ..., Id: Exp }
    Record(Record),
    /// Exp.Id
    FieldAccess(ExpBox, Id),
    /// null
    Null,
    /// Error
    Error,
}

impl From<i64> for Exp {
    fn from(i: i64) -> Self {
        Exp::Int(i)
    }
}

impl From<Id> for Exp {
    fn from(id: Id) -> Self {
        Exp::Id(id)
    }
}

impl From<&str> for Exp {
    fn from(s: &str) -> Self {
        Exp::Id(Id::from(s))
    }
}

pub type StmBox = Box<Stm>;

#[derive_common]
#[derive(SuperFoldable)]
/// Statement
pub enum Stm {
    /// Id = Exp;
    Assign(Id, Exp),
    /// output Exp;
    Output(Exp),
    /// Stm*
    Block(Vec<Stm>),
    /// if Exp { Stm } else { Stm }
    If(Exp, StmBox, Option<StmBox>),
    /// while Exp { Stm }
    While(Exp, StmBox),
    /// var Id (, Id)*;
    Locals(Vec<Id>),
    /// return Exp;
    Return(Exp),
    /// *Exp = Exp;
    Store(Exp, Exp),
    /// Id.Id = Exp;
    FieldUpdate(Id, Id, Exp),
    /// (*Exp).Id = Exp;
    IndirectFieldUpdate(Exp, Id, Exp),
}

#[derive_common]
#[derive(SuperFoldable)]
/// Function
///
/// name(arg, arg, arg) { locals? stm return exp }
pub struct Fun {
    pub name: Id,
    pub args: Vec<Id>,
    pub locals: Vec<Id>,
    pub body: Vec<Stm>,
    pub ret: Exp,
}

#[derive_common]
#[derive(SuperFoldable)]
pub struct Program {
    pub funs: Vec<Fun>,
}

#[derive_common]
#[derive(Foldable)]
pub struct Record {
    pub fields: Vec<(Id, Exp)>,
}

impl From<Vec<(Id, Exp)>> for Record {
    fn from(fields: Vec<(Id, Exp)>) -> Self {
        Record { fields }
    }
}
