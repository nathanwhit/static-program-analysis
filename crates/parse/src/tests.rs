mod sourcegen_inline_tests;

use std::{
    fs,
    path::{Path, PathBuf},
};

use ast::{AstCtx, Exp, Fun, Program, Stm};
use insta::{assert_display_snapshot, assert_ron_snapshot};

use crate::{ParseResult, Parser};

#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug)]
enum EntryPoint {
    SourceFile,
    Exp,
    Stm,
    Fun,
}

enum Ast {
    Exp(Exp),
    Stm(Stm),
    Fun(Fun),
    Program(Program),
}

impl EntryPoint {
    pub fn parse<'i>(&self, input: &'i str) -> (ParseResult<Ast>, AstCtx) {
        let parse: fn(&mut Parser<'_>) -> ParseResult<Ast> = match self {
            EntryPoint::Exp => |p| crate::grammar::exp(p).map(Ast::Exp),
            EntryPoint::Stm => |p| crate::grammar::stm(p).map(Ast::Stm),
            EntryPoint::SourceFile => |p| crate::grammar::program(p).map(Ast::Program),
            EntryPoint::Fun => |p| crate::grammar::fun(p).map(Ast::Fun),
        };
        let mut p = Parser::new(input).unwrap();
        let res = parse(&mut p);
        if res.is_ok() {
            assert!(
                p.at_end(),
                "parser did not consume entire input. on {:?}",
                p.current_tok()
            );
        }
        (res, p.finish())
    }
}

// Uncomment if we add any of these tests
// #[test]
// fn lex_ok() {
//     insta::glob!("../test_data", "lexer/ok/*.tip", |path| {
//         let case = TestCase::read(path);
//         let actual = lex(&case.text);
//         insta::with_settings!({description => &case.text, omit_expression => true }, { assert_snapshot!(actual) });
//     });
// }

// Uncomment if we add any of these tests
// #[test]
// fn lex_err() {
//     for case in TestCase::list("lexer/err") {
//         let actual = lex(&case.text);
//         assert_snapshot!(actual)
//     }
// }

macro_rules! test_glob {
    (@helper $actual: ident, $ctx: ident, $case: ident) => {
        {
            let wrapped = common::serde::SerializeWrapper::new(&$actual, &$ctx);
            insta::with_settings!({description => &$case.text, omit_expression => true, sort_maps => true }, { assert_ron_snapshot!(wrapped) });
        }
    };
    (ok $glob: expr) => {
        insta::glob!("../test_data", &format!("parse/{}/*.tip", $glob), |path| {
            let case = TestCase::read(path);
            eprintln!("running test: {}", case.tip.display());
            let (actual, ctx) = case.entry.parse(&case.text);
            let actual = actual.unwrap();
            match actual {
                Ast::Exp(actual) => test_glob!(@helper actual, ctx, case),
                Ast::Stm(actual) => test_glob!(@helper actual, ctx, case),
                Ast::Fun(actual) => test_glob!(@helper actual, ctx, case),
                Ast::Program(actual) => test_glob!(@helper actual, ctx, case),
            }
        });
    };
    (err $glob: expr) => {
        insta::glob!("../test_data", &format!("parse/{}/*.tip", $glob), |path| {
            let case = TestCase::read(path);
            eprintln!("running test: {}", case.tip.display());
            let (actual, _ctx) = case.entry.parse(&case.text);
            let actual = actual.map(|_| ()).unwrap_err();

            insta::with_settings!({description => &case.text, omit_expression => true, sort_maps => true}, { assert_display_snapshot!(actual) });
        });
    };
}

// #[test]
// fn parse_ok() {
//     test_glob!(ok "ok/file");
// }

// Uncomment if we add any of these tests
// #[test]
// fn parse_err() {
//     for case in TestCase::list("parser/err") {
//         let (actual, errors) = parse(EntryPoint::SourceFile, &case.text);
//         assert!(
//             errors,
//             "no errors in an ERR file {}:\n{actual}",
//             case.tip.display()
//         );
//         assert_snapshot!(actual)
//     }
// }

#[test]
fn parse_inline_ok() {
    test_glob!(ok "inline/ok/exp");
    test_glob!(ok "inline/ok/stm");
    test_glob!(ok "inline/ok/fun");
    test_glob!(ok "inline/ok/file");
}

#[test]
fn parse_inline_err() {
    test_glob!(err "inline/err/exp");
    test_glob!(err "inline/err/stm");
    test_glob!(err "inline/err/fun");
    // test_glob!(err "inline/err/file");
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Debug)]
struct TestCase {
    tip: PathBuf,
    entry: EntryPoint,
    text: String,
}

fn entry_point_from_str(s: &str) -> EntryPoint {
    match s {
        "exp" => EntryPoint::Exp.into(),
        "file" => EntryPoint::SourceFile.into(),
        "stm" => EntryPoint::Stm.into(),
        "fun" => EntryPoint::Fun.into(),
        "lexer" => todo!(),
        _ => panic!("unknown entry point {s}"),
    }
}

impl TestCase {
    fn read(path: &Path) -> TestCase {
        if path
            .parent()
            .unwrap()
            .parent()
            .unwrap()
            .file_name()
            .unwrap()
            == "lexer"
        {
            todo!()
        }
        let entry = entry_point_from_str(
            path.parent()
                .unwrap()
                .file_name()
                .unwrap()
                .to_str()
                .unwrap(),
        );
        if path.extension().unwrap_or_default() == "tip" {
            let tip = path.into();
            let text = fs::read_to_string(&tip).unwrap();
            TestCase { tip, text, entry }
        } else {
            panic!("unknown file extension {}", path.display());
        }
    }

    #[allow(unused)]
    fn list(path: &str) -> Vec<TestCase> {
        let crate_root_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
        let test_data_dir = crate_root_dir.join("test_data");
        let dir = test_data_dir.join(path);

        let mut res = Vec::new();
        let read_dir = fs::read_dir(&dir)
            .unwrap_or_else(|err| panic!("can't `read_dir` {}: {err}", dir.display()));
        for file in read_dir {
            let file = file.unwrap();
            let path = file.path();
            if path.is_dir() {
                let entry = entry_point_from_str(path.file_name().unwrap().to_str().unwrap());
                for case in Self::list(path.to_str().unwrap()) {
                    res.push(TestCase { entry, ..case })
                }
                continue;
            }
            if path.extension().unwrap_or_default() == "tip" {
                let tip = path;
                let text = fs::read_to_string(&tip).unwrap();
                res.push(TestCase {
                    tip,
                    text,
                    entry: EntryPoint::SourceFile.into(),
                });
            }
        }
        res.sort();
        res
    }
}
