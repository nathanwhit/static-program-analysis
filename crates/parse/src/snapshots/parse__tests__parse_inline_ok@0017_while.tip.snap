---
source: crates/parse/src/tests.rs
description: "while 1 { a = 1; }\n"
input_file: crates/parse/test_data/parse/inline/ok/stm/0017_while.tip
---
While(Int(1), Block([
  Assign(Id("a"), Int(1)),
]))
